docs/net.md

```diff
@@ -0,0 +1,343 @@
+            }
+            Err(ref e) if e.kind() == WouldBlock => {
+                // The resource is not ready, fall through to task registration
+            }
+            Err(e) => {
+                // All other errors are returned to the caller
+                return Err(e);
+            }
+        }
+    }
+
+    // The task is set even if it is already `Some`, this handles the case where
+    // the resource is moved to a different task than the one stored in
+    // `self.task`.
+    registration.task = Some(task::current());
+    Ok(Async::NotReady)
+}
+```
+
+Note that there is only a single `task` field per resource. The implications are
+that a resource can only be used from a single task at a time. If
+`TcpListener::poll_accept` returns `NotReady`, registering the current task and
+the listener is then sent to a different task which calls `poll_accept` and sees
+`NotReady`, then the second task is the only one that will receive a
+notification once a socket is ready to be accepted. Resources may support
+tracking different tasks for different operations. For example, `TcpStream`
+internally has two task fields: one for notifying on read ready and one for
+notifying on write ready. This allows `TcpStream::poll_read` and
+`TcpStream::poll_write` to be called from different tasks.
+
+The evented types are registered with the driver's mio::Poll instance as part of
+the `register` function used above (again, this guide uses a **simplified**
+implementation which does not match the actual one in `tokio-reactor` but is
+sufficient for understanding how `tokio-reactor` behaves).
+
+```rust
+fn register<T: mio::Evented>(&mut self, evented: &T) -> Arc<Mutex<Registration>> {
+    // Generate a unique identifier for this registration. This identifier
+    // can be converted to and from a Mio Token.
+    let id = generate_unique_identifier();
+
+    // Register the I/O type with Mio
+    self.poll.register(
+        evented, id.into_token(),
+        mio::Ready::all(),
+        mio::PollOpt::edge());
+
+    let registration = Arc::new(Mutex::new(Registration {
+        id,
+        task: None,
+    }));
+
+    self.resources.insert(id, registration.clone());
+
+    registration
+}
+```
+
+### Running the driver
+
+The driver needs to run in order for its associated resources to function. If
+the driver does not run, the resources will never become ready. Running the
+driver is handled automatically when using a Runtime, but it is useful to
+understand how it works. If you are interested in the real implementation, the
+`tokio-reactor` source is the best reference.
+
+When resources are registered with the driver, they are also registered with
+Mio. Running the driver performs the following steps in a loop:
+
+1) Call `Poll::poll` to get operating system events.
+2) Dispatch all events to the appropriate resources via the registration.
+
+The steps above are done by calling `Reactor::turn`. The looping part is up to
```

Comment:
Maybe mention that the driver is typically running in a background thread that loops and calls `turn` until the runtime is shut down? Otherwise, the question is: who is supposed to run this loop?